src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs

```diff
@@ -1131,6 +1150,56 @@ public static void SetPrimitiveValueConverter(this IEdmModel model, IEdmTypeDefi
             model.SetPrimitiveValueConverter(typeDefinition.Definition, converter);
         }
 
+        /// <summary>
+        /// Marks the model as immutable. This may enable optimizations
+        /// that assume the model does not get modified. It is the responsibility
+        /// of the caller to ensure that the underlying EDM model is not modified after calling
+        /// this method.
+        /// </summary>
+        /// <param name="model"></param>
+        public static void MarkAsImmutable(this IEdmModel model)
+        {
+            EdmUtil.CheckArgumentNull(model, "model");
+
+            // it doesn't matter what value we store, so long as it's not null
+            model.SetAnnotationValue(model, EdmConstants.InternalUri, CsdlConstants.IsImmutable, new object());

```

Comment:
There are some static objects in `CacheHelper` like `CacheHelper.Unknown`, `CacheHelper.BoxedBool(true/false)` which I initially thought would be good candidates to use here to avoid allocating a new object. But I'm not sure if they were intendent for that purpose. Also not sure if they would make a significant difference since this object is only allocated once for the entire lifetime of the model.